core/contracts/Governor.sol

```diff
@@ -0,0 +1,139 @@
+
+import "./Finder.sol";
+import "./MultiRole.sol";
+import "./FixedPoint.sol";
+import "./Voting.sol";
+import "./Testable.sol";
+
+
+/**
+ * @title Takes proposals for certain governance actions and allows UMA token holders to vote on them.
+ */
+contract Governor is MultiRole, Testable {
+
+    enum Roles {
+        // Can set the proposer.
+        Admin,
+        // Address that can make proposals.
+        Proposer
+    }
+
+    struct Proposal {
+        address to;
+        uint value;
+        bytes data;
+        uint requestTime;
+    }
+
+    Finder internal finder;
+    Proposal[] public proposals;
+
+    /**
+     * @notice Emitted when a new proposal is created.
+     */
+    event NewProposal(uint indexed id, address indexed to, uint value, bytes data);
+
+    /**
+     * @notice Emitted when an existing proposal is executed.
+     */
+    event ProposalExecuted(uint indexed id);
+
+    constructor(address _finderAddress, bool _isTest) public Testable(_isTest) {
+        finder = Finder(_finderAddress);
+        _createExclusiveRole(uint(Roles.Admin), uint(Roles.Admin), msg.sender);
+        _createExclusiveRole(uint(Roles.Proposer), uint(Roles.Admin), msg.sender);
+    }
+
+    /**
+     * @notice Proposes a new governance action. Can only be called by the holder of the Proposer role.
+     * @param to the address to call.
+     * @param value the ETH value to attach to the call.
+     * @param data the transaction data to attach to the call.
+     * @dev You can create the data portion of the transaction by doing the following:
+     * ```
+     * const truffleContractInstance = await TruffleContract.deployed()
+     * const data = truffleContractInstance.methods.methodToCall(arg1, arg2).encodeABI()
+     * ```
+     */
+    function propose(address to, uint value, bytes calldata data) external onlyRoleHolder(uint(Roles.Proposer)) {
+        uint id = proposals.length;
+        uint time = getCurrentTime();
+
+        proposals.push(Proposal({
+            to: to,
+            value: value,
+            data: data,
+            requestTime: time
+        }));
+
+        bytes32 identifier = _constructIdentifier(id);
+
+        // Request a vote on this proposal in the DVM.
+        Voting voting = _getVoting();
+        voting.addSupportedIdentifier(identifier);
+        voting.requestPrice(identifier, time);
+        voting.removeSupportedIdentifier(identifier);
+
+        emit NewProposal(id, to, value, data);
+    }
+
+    /**
+     * @notice Executes a proposed governance action that has been approved by voters. This can be called by anyone.
+     */
+    function executeProposal(uint id) external {
+        Proposal storage proposal = proposals[id];
+        int price = _getVoting().getPrice(_constructIdentifier(id), proposal.requestTime);
+
+        require(price != 0, "Cannot execute, proposal was voted down.");
+        require(_executeCall(proposal.to, proposal.value, proposal.data), "Proposal execution failed.");
```

Comment:
Just to double check, this method can only be successfully once per proposal?

Reply:
That's how it's supposed to work, but I don't think it prevents multiple calls right now. Nice catch.

Reply:
Done.